kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,72 @@
+//! Interface for Digest
+
+use crate::common::leasable_buffer::LeasableBuffer;
+use crate::returncode::ReturnCode;
+
+/// The 'types' of digests, this should define the output size of the digest
+/// operations.
+pub trait DigestType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl DigestType for [u8; 32] {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a, T: DigestType> {
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// On error or success `data` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a digest is computed.
+    /// On error or success `digest` will contain a reference to the original
+    /// digest supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), ReturnCode>, digest: &'static mut T);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a, T: DigestType> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a, T>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
+    fn add_data(
```

Comment:
The name of the method as well as the documentation lead me to think that if the data I want to hash doesn't fit in RAM entirely, I can call repeatedly the `add_data` function over several chunks (waiting for the callback between every chunk of course) and then call `run` to get the digest.
If this is the case, then I think I'm missing a `start_new_message` method or something similar. Otherwise the digest will never be correct because the initial value will reloaded at each chunk. If I read the code correctly that's what the current implementation for OT module is doing in this PR:
```rust
regs.cmd.modify(CMD::START::SET);
```

Maybe I'm too used to standard crypto hashing API (`new`, `update`, `finish`).

Reply:
The documentation and name are correct and you can call `add_data()` multiple times to hash something that doesn't fit into RAM.

Why do you need a `start_new_message()` function?

Reply:
Ok so if I got it correctly, the `start_new_message()` function I was looking for is actually the`set_mode_hmacsha256()` function. That's how you tell the hardware module that you're going to compute another hash or hmac.
Let's say I need to compute hmac-sha256 with the same key for 2 message, with your HIL, am I supposed to do?

1. `set_mode_hmacsha256(key)`
1. `add_data(msg1_part1)`, `add_data(msg1_part2)`, ...., `add_data(msg1_partN)`
1. `run(mac_msg1)`
1. `set_mode_hmacsha256(key)`
1. `add_data(msg1_part1)`, `add_data(msg1_part2)`, ...., `add_data(msg1_partN)`
1. `run(mac_msg1)`

or is the step number 4 optional?

I'm trying to implement your HIL for the ARM CryptoCell-31x and it requires manually load the initial value for the selected digest algorithm. That's why I'm trying to figure out when to do it. If it is in the `set_mode_hmacsha256()` function, then we have a 1:1 equivalent to the more traditional `new()` (`set_mode_hmacsha256`), `update()` (`add_data()`), `finish()` (`run()`) API I've been mentioning.
If we don't need to call `set_mode_hmacsha256()` between two messages, I need to implement it both in `set_mode_hmacsha256()` and in `run()` functions.

Hope this makes sense :)

Reply:
Step 4 should be optional in this case (untested though).

I'm happy to update the documentation and implementation to make step 4 compulsory if required.